File: sdk/python/iploop/support.py
"""Support API client."""

import sys
import requests
from .exceptions import AuthError, QuotaExceeded


class SupportClient:
    def __init__(self, api_key, api_base):
        self.api_key = api_key
        self.api_base = api_base.rstrip("/")
        self._headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}

    def _get(self, path):
        resp = requests.get(f"{self.api_base}{path}", headers=self._headers, timeout=15, verify=True)
        if resp.status_code == 401:
            raise AuthError("Invalid API key")
        resp.raise_for_status()
        return resp.json()

    def _post(self, path, json_data):
        resp = requests.post(f"{self.api_base}{path}", headers=self._headers, json=json_data, timeout=15, verify=True)
        if resp.status_code == 401:
            raise AuthError("Invalid API key")
        resp.raise_for_status()
        return resp.json()

    def usage(self):
        data = self._get("/api/support/diagnose")
        self._check_quota(data)
        return data

    def status(self):
        return self._get("/api/support/status")

    def ask(self, question):
        return self._post("/api/support/ask", {"question": question})

    def countries(self):
        return self._get("/api/support/countries")

    def _check_quota(self, data):
        try:
            used = data.get("used_gb", 0)
            total = used + data.get("remaining_gb", 999)
            if total > 0:
                pct = used / total * 100
                if pct >= 100:
                    raise QuotaExceeded()
                if pct >= 80:
                    print(
                        f"⚠️  IPLoop: {pct:.0f}% bandwidth used. Upgrade at https://iploop.io/pricing",
                        file=sys.stderr,
                    )
        except (TypeError, KeyError):
            pass

File: sdk/node/src/support.js
const https = require('https');
const { AuthError, QuotaExceeded } = require('./exceptions');

class SupportClient {
  constructor(apiKey, apiBase) {
    this.apiKey = apiKey;
    this.apiBase = apiBase.replace(/\/+$/, '');
    this.headers = { 'Authorization': `Bearer ${apiKey}`, 'Content-Type': 'application/json' };
  }

  _request(method, path, body) {
    return new Promise((resolve, reject) => {
      const url = new URL(this.apiBase + path);
      const opts = {
        hostname: url.hostname, port: url.port || 443, path: url.pathname,
        method, headers: { ...this.headers }, timeout: 15000,
        rejectUnauthorized: true,
      };
      if (body) {
        const data = JSON.stringify(body);
        opts.headers['Content-Length'] = Buffer.byteLength(data);
      }
      const req = https.request(opts, res => {
        let chunks = [];
        res.on('data', c => chunks.push(c));
        res.on('end', () => {
          const text = Buffer.concat(chunks).toString();
          if (res.statusCode === 401) return reject(new AuthError());
          if (res.statusCode >= 400) return reject(new Error(`HTTP ${res.statusCode}: ${text}`));
          try { resolve(JSON.parse(text)); } catch { resolve(text); }
        });
      });
      req.on('error', reject);
      req.on('timeout', () => { req.destroy(); reject(new Error('Request timeout')); });
      if (body) req.write(JSON.stringify(body));
      req.end();
    });
  }

  async usage() {
    const data = await this._request('GET', '/api/support/diagnose');
    this._checkQuota(data);
    return data;
  }
  status() { return this._request('GET', '/api/support/status'); }
  ask(question) { return this._request('POST', '/api/support/ask', { question }); }
  countries() { return this._request('GET', '/api/support/countries'); }

  _checkQuota(data) {
    try {
      const used = data.used_gb || 0;
      const total = used + (data.remaining_gb || 999);
      if (total > 0) {
        const pct = used / total * 100;
        if (pct >= 100) throw new QuotaExceeded();
        if (pct >= 80) process.stderr.write(`⚠️  IPLoop: ${pct.toFixed(0)}% bandwidth used. Upgrade at https://iploop.io/pricing\n`);
      }
    } catch (e) { if (e instanceof QuotaExceeded) throw e; }
  }
}
module.exports = { SupportClient };
